### PR TITLE
Feature action button layout #13

### PR DESCRIPTION
**Pull request by Blackbaud-BrandonJones**
first:
screen shot 2017-04-04 at 4 50 32 pm

second:
screen shot 2017-04-04 at 4 50 43 pm

### CODE REVIEW

File: src/modules/action-button/action-button-link.ts [hunk 1]
```diff
@@ -0,0 +1,6 @@
+import { StacheNavLink } from '../nav/nav-link';
+
+export interface StacheActionButtonLink extends StacheNavLink {
```
Comment (Blackbaud-SteveBrush) [inline src/modules/action-button/action-button-link.ts]: While I think this approach is best for the Action Button component, we may be missing an opportunity to use the icon and summary for other components that may require it. For example, we may (at some point) want to create another nav component that utilizes an icon or a summary. It might be better to adjust the `StacheNavLink` interface, instead of creating a new one for this component.

```
export interface StacheNavLink {
  name: string;
  path: string[] | string;
  fragment?: string;
  icon?: string;
  summary?: string;
}
```

I could also see a use for this on the breadcrumbs component: changing "Home" to use a house icon...

File: src/modules/action-button/action-button.component.scss [hunk 2]
```diff
@@ -0,0 +1,14 @@
+/deep/
+.stache-action-buttons {
+    display: inline-flex;
+    flex-wrap: wrap;
+
+  sky-action-button {
```
Comment (Blackbaud-SteveBrush) [inline src/modules/action-button/action-button.component.scss]: Styles on classnames are (albeit negligibly) faster than element names. Just good practice to style against classnames when possible. http://stackoverflow.com/questions/39111878/which-css-selector-is-faster-in-performance-class-in-class-vs-tagname-in-class

File: src/modules/action-button/action-button.component.scss [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+    display: inline-flex;
+  }
+  
+  .sky-action-button {
+    width: 236px;
```
Comment (Blackbaud-SteveBrush) [inline src/modules/action-button/action-button.component.scss]: Specific width? Why not flex?

File: src/modules/layout/layout.component.html [hunk 4]
```diff
@@ -41,4 +41,16 @@
   </stache-layout-sidebar>
 </template>
 
+<template #actionButtonLayout>
+  <stache-layout-action-button
+    pageTitle="{{pageTitle}}"
+    [showBackToTop]="showBackToTop"
+    [showBreadcrumbs]="showBreadcrumbs"
+    [breadcrumbsRoutes]="breadcrumbsRoutes"
+    [actionButtonRoutes]="actionButtonRoutes">
```
Comment (Blackbaud-SteveBrush) [inline src/modules/layout/layout.component.html]: Seeing this here, it's obvious that we need to rethink how we're passing navigation routes to each layout component. As I see it, there are only two types of layout routes: routes used for inter-page navigation (URLs) and routes used for intra-page navigation (page anchors). 

Inter-page navigation could include sidebar, action buttons, panels, whatever.

Intra-page navigation would be any page anchors for the current page.

We can leave it as `actionButtonRoutes` for this PR, but let's think about what would be the best name for these two categories for a new PR. (Perhaps we need the assistance of a tech writer: @blackbaud-brandonhare.)

File: src/modules/action-button/action-button.component.scss (hunk 3 above)
Comment (Blackbaud-BrandonJones) [inline src/modules/action-button/action-button.component.scss]: Having the cards different widths made them look pretty terrible and non-uniform,  I can experiment a bit more.

File: src/modules/action-button/action-button-link.ts (hunk 1 above)
Comment (Blackbaud-SteveBrush) [inline src/modules/action-button/action-button-link.ts]: Bump.

Comment (blackbaud-brandonhare) [inline src/modules/action-button/action-button-link.ts]: Ba bump ba bump bump.

File: src/modules/breadcrumbs/breadcrumbs.component.scss [hunk 5]
```diff
@@ -2,35 +2,36 @@
 
 /deep/ 
 .stache-breadcrumbs {
```
Comment (Blackbaud-SteveBrush) [inline src/modules/breadcrumbs/breadcrumbs.component.scss]: This is a good fix, but let's pull it out into its own pull request.

File: src/modules/back-to-top/back-to-top.component.spec.ts [hunk 6]
```diff
@@ -24,7 +24,7 @@ describe('StacheBackToTopComponent', () => {
     debugElement = fixture.debugElement;
   });
 
-  beforeEach(inject([WindowRefService], service => {
+  beforeEach(inject([WindowRefService], (service: any) => {
```
Comment (blackbaud-brandonhare) [inline src/modules/back-to-top/back-to-top.component.spec.ts]: Barf. Typescript

Comment (Blackbaud-SteveBrush) [inline src/modules/back-to-top/back-to-top.component.spec.ts]: Let's pull this out into a hotfix branch (or make @blackbaud-brandonhare fix it).

Comment (blackbaud-brandonhare) [inline src/modules/back-to-top/back-to-top.component.spec.ts]: FINE. I WILL FIX IT. ARE YOU HAPPY STEVE?

File: src/modules/action-button/action-button.component.scss [hunk 7]
```diff
@@ -0,0 +1,29 @@
+/deep/
+.stache-action-buttons {
+    display: inline-flex;
+    flex-wrap: wrap;
+
+  sky-action-button {
+    display: inline-flex;
+    flex: 0 1 20%;
+    @media (max-width: 992px) {
```
Comment (Blackbaud-SteveBrush) [inline src/modules/action-button/action-button.component.scss]: Did you already attempt to use `StacheColumnComponent` for the breakpoints, in your HTML template? It'd be nice if we could keep the media query "logic" isolated.

File: src/modules/back-to-top/back-to-top.component.spec.ts (hunk 6 above)
Comment (Blackbaud-SteveBrush) [inline src/modules/back-to-top/back-to-top.component.spec.ts]: Actually, yes, yes I am.

File: src/modules/action-button/action-button.component.html [hunk 8]
```diff
@@ -0,0 +1,17 @@
+<div class="stache-action-buttons" *ngIf="actionButtonRoutes">
+  <stache-column *ngFor="let route of actionButtonRoutes" screenSmall="6" screenMedium="4" screenLarge="3">
```
Comment (Blackbaud-SteveBrush) [inline src/modules/action-button/action-button.component.html]: Also needs to be wrapped in a `<stache-row>` :)

File: src/modules/action-button/action-button.component.spec.ts [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+      ],
+      providers: [{ provide: Router, useValue: mockRouter }]
+    })
+    .compileComponents();
+    fixture = TestBed.createComponent(StacheActionButtonComponent);
+    component = fixture.componentInstance;
+    debugElement = fixture.debugElement;
+  });
+
+  it('Should create an action button component.', () => {
```
Comment (Blackbaud-SteveBrush) [inline src/modules/action-button/action-button.component.spec.ts]: What's this `it` doing outside of a `describe`?? We probably don't even need this "test". :)

File: src/modules/action-button/action-button.component.spec.ts [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+    expect(component).toExist();
+  });
+
+  describe('navigate', () => {
```
Comment (Blackbaud-SteveBrush) [inline src/modules/action-button/action-button.component.spec.ts]: Put the name of the component, here (I had to correct this in one of my branches, too). 
```
describe('StacheActionButtonComponent', () => {});
```

File: src/modules/action-button/action-button.component.spec.ts [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+    it('Should navigate to the new link.', () => {
```
Comment (Blackbaud-SteveBrush) [inline src/modules/action-button/action-button.component.spec.ts]: Lower-case "should", please. Create a sentence, "StacheActionButtonComponent should navigate to the new link."

File: src/modules/action-button/action-button.component.spec.ts [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+      expect(mockRouter.navigate).toHaveBeenCalledWith(['/home'], {});
+    });
+
+    it('Should navigate given a fragment with the route.', () => {
+      let newRoute = { name: 'home', path: '/home', fragment: 'section-two' };
+      component.navigate(newRoute);
+      expect(mockRouter.navigate).toHaveBeenCalledWith(['/home'], { fragment: 'section-two' });
+    });
+
+    it('Should navigate given a path from an array of paths.', () => {
```
Comment (Blackbaud-SteveBrush) [inline src/modules/action-button/action-button.component.spec.ts]: Great tests, overall!

File: src/modules/action-button/action-button.component.spec.ts (hunk 10 above)
Comment (Blackbaud-BrandonJones) [inline src/modules/action-button/action-button.component.spec.ts]: screen shot 2017-04-10 at 5 05 08 pm

Is nesting describe blocks inside the component to break down functionality not appropriate?

File: src/modules/action-button/action-button.component.spec.ts (hunk 9 above)
Comment (Blackbaud-BrandonJones) [inline src/modules/action-button/action-button.component.spec.ts]: This is inside the base describe block, the `describe('Action Button Component', ( ) +> {`

File: src/modules/layout/layout.component.html (hunk 4 above)
Comment (blackbaud-brandonhare) [inline src/modules/layout/layout.component.html]: Agreed. The amount of routes we are passing in is getting crazy. We should come up with a more maintainable pattern.

**Comment by Blackbaud-SteveBrush** [comment]
I'm having a difficult time bringing all these files together in my mind. I know it's a pain, but can you please create two separate branches, one for the component (including the test and demo), and one for the layout? I think it will help to have them separated so that we can talk about them in specific terms, as both are deceptively complex and add a lot to Stache's offerings.

Also, naming the component `StacheActionButtonComponent` falsely conveys two things: 1) we created our own version of Action Button, and 2) this component represents a singular "thing". I think changing the name to `StacheActionButtonListComponent` better describes its purpose. I admit it's wordy, but there's less confusion as to what it is, and what it does. I'm welcome to other ideas for the name.

The `navigate` method is nearly a copy/paste from `StacheNavComponent`. This might be a situation where [component inheritance](https://scotch.io/tutorials/component-inheritance-in-angular-2) is appropriate. Worth researching!

File: src/modules/breadcrumbs/breadcrumbs.component.scss (hunk 5 above)
Comment (Blackbaud-SteveBrush) [inline src/modules/breadcrumbs/breadcrumbs.component.scss]: 👍

File: src/app/app.scss [hunk 13]
```diff
@@ -6,6 +6,6 @@
   .stache-column {
     background-color: lighten($sky-color-aqua, 30%);
     border: 1px solid lighten($sky-color-aqua, 15%);
-    word-break: break-all;
+    overflow-wrap: break-word;
```
Comment (Blackbaud-SteveBrush) [inline src/app/app.scss]: Why the change? Was it giving you problems?

File: src/modules/layout/layout-action-buttons.component.html [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+          <stache-page-title>
+            {{pageTitle}}
+          </stache-page-title>
+        </stache-page-header>
+
+        <ng-content></ng-content>
+      </stache-column>
+    </stache-row>
+
+  <div class="stache-action-buttons" *ngIf="actionButtonRoutes">
```
Comment (Blackbaud-SteveBrush) [inline src/modules/layout/layout-action-buttons.component.html]: Double-check your tabs in this file...

File: src/modules/layout/layout.component.ts [hunk 15]
```diff
@@ -43,6 +46,9 @@ export class StacheLayoutComponent implements OnInit, StacheLayout {
   @ViewChild('sidebarLayout')
   private sidebarTemplateRef: any;
 
+    @ViewChild('actionButtonsLayout')
+  private actionButtonsTemplateRef: any;
```
Comment (Blackbaud-SteveBrush) [inline src/modules/layout/layout.component.ts]: Double-check your tabs in this file...

File: src/app/app.scss (hunk 13 above)
Comment (Blackbaud-BrandonJones) [inline src/app/app.scss]: screen shot 2017-04-11 at 10 28 28 am
screen shot 2017-04-11 at 10 29 43 am

File: src/modules/layout/layout-action-buttons.component.ts [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+import { Router } from '@angular/router';
+import { StacheLayout } from './layout';
+import { InputConverter } from '../shared';
+
+@Component({
+  selector: 'stache-layout-action-buttons',
+  templateUrl: './layout-action-buttons.component.html',
+  styleUrls: ['./layout-action-buttons.component.scss']
+})
+export class StacheLayoutActionButtonsComponent implements StacheLayout {
```
Comment (Blackbaud-SteveBrush) [inline src/modules/layout/layout-action-buttons.component.ts]: Try this:
```
export class StacheLayoutActionButtonsComponent extends StacheNavComponent implements StacheLayout {
```
Remove the `constructor`, the `Router` import, and the `navigate` method.

File: src/modules/layout/layout-action-buttons.component.ts [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+  public pageTitle: string;
+
+  @Input()
+  public breadcrumbsRoutes: StacheNavLink[];
+
+  @Input()
+  public inPageRoutes: StacheNavLink[];
+
+  @Input()
+  public actionButtonRoutes: StacheNavLink[];
```
Comment (Blackbaud-SteveBrush) [inline src/modules/layout/layout-action-buttons.component.ts]: Remove this, and change to `routes` in the layout template? Since we're extending the nav component...
